lib/streamlit/connections/snowflake_connection.py

```diff
@@ -0,0 +1,252 @@
+# way to configure this at a per-line level :(
+# mypy: no-warn-unused-ignores
+
+from __future__ import annotations
+
+from datetime import timedelta
+from typing import TYPE_CHECKING, Any, cast
+
+import pandas as pd
+
+from streamlit.connections import BaseConnection
+from streamlit.connections.util import running_in_sis
+from streamlit.errors import StreamlitAPIException
+from streamlit.runtime.caching import cache_data
+
+if TYPE_CHECKING:
+    from snowflake.connector import (  # type:ignore[import] # isort: skip
+        SnowflakeConnection as InternalSnowflakeConnection,
+    )
+    from snowflake.connector.cursor import SnowflakeCursor  # type:ignore[import]
+    from snowflake.snowpark.session import Session  # type:ignore[import]
+
+
+_REQUIRED_CONNECTION_PARAMS = {"account"}
+
+
+def _validate_connection_params(required: set[str], params: dict[str, Any]) -> None:
+    missing = required - set(params)
+    if len(missing) > 0:
+        raise StreamlitAPIException(
+            f"Missing Snowflake connection params: {', '.join(missing)}"
+        )
+
+
+class SnowflakeConnection(BaseConnection["InternalSnowflakeConnection"]):
+    """A connection to Snowflake using the Snowflake Python Connector. Initialize using
+    ``st.connection("<name>", type="snowflake")``.
+
+    SnowflakeConnection supports direct SQL querying using ``.query("...")``, access to
+    the underlying Snowflake Python Connector object with ``.raw_connection``, and other
+    convenience functions. See the methods below for more information.
+    SnowflakeConnections should always be created using ``st.connection()``, **not**
+    initialized directly.
+    """
+
+    def _connect(self, **kwargs) -> "InternalSnowflakeConnection":
+        import snowflake.connector  # type:ignore[import]
+        from snowflake.snowpark.context import get_active_session  # type:ignore[import]
+
+        # If we're running in SiS, just call get_active_session() and retrieve the
+        # lower-level connection from it.
+        if running_in_sis():
+            session = get_active_session()
+
+            if hasattr(session, "connection"):
+                return session.connection
+            return session._conn._conn
+
+        # We require qmark-style parameters everywhere for consistency across different
+        # environments where SnowflakeConnections may be used.
+        snowflake.connector.paramstyle = "qmark"
+
+        # Otherwise, attempt to create a new connection from whatever credentials we
+        # have available.
+        st_secrets = self._secrets.to_dict()
+        if len(st_secrets):
+            conn_kwargs = {**st_secrets, **kwargs}
+            _validate_connection_params(_REQUIRED_CONNECTION_PARAMS, conn_kwargs)
+            return snowflake.connector.connect(**conn_kwargs)
+
+        if hasattr(snowflake.connector.connection, "CONFIG_MANAGER"):
+            config_mgr = snowflake.connector.connection.CONFIG_MANAGER
```

Comment:
nit: is the `CONFIG_MANAGER` also something that was added in more recent versions of the snowflake connector?

Reply:
Yeah -- this is only available in more recent versions -- I'll add an explanatory comment in a followup PR